crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -26,12 +29,126 @@ use crate::{
 use std::{cmp::Ordering, collections::HashMap, ops::Deref, rc::Rc};
 use vec1::Vec1;
 
+pub(crate) fn infer_function(
+    fun: &UntypedFunction,
+    module_name: &str,
+    hydrators: &mut HashMap<String, Hydrator>,
+    environment: &mut Environment<'_>,
+    lines: &LineNumbers,
+    tracing: Tracing,
+) -> Result<Function<Rc<Type>, TypedExpr, TypedArg>, Error> {
+    if let Some(typed_fun) = environment.inferred_functions.get(&fun.name) {
+        return Ok(typed_fun.clone());
+    };
```

Comment:
Some memoization happening here, now-necessary since we might have already processed that definition. This is because:

1. There might be multiple function calls to that function
2. Simply because it's processed as part of the normal "infer_definition" loop, but has already been seen in a call from a previous definition.